### PR TITLE
Coding guidelines

### PR DESCRIPTION
**Pull request by otoolep**

### CODE REVIEW

File: CODING_GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+# The Guidelines
+
+## Try not to use third-party libaries
+
+A third-party package is defined as one that is not part of the standard Go distribution. Generally speaking we prefer to minimize our use of third-party packages, and avoid them unless absolutely necessarily. We'll often write a little bit of code rather than pull in a third-party package. Of course, we do use some third-party packages -- most importantly we use [BoltDB](https://github.com/boltdb/bolt) in some storage engines. So to maximise the chance your change will be accepted by us, use only the standard libraries, or the third-party packages we have decided to use.
+
+For rationale, check out the post [The Case Against Third Party Libraries](http://blog.gopheracademy.com/advent-2014/case-against-3pl/).
+
+## Always include a default case in a 'switch' statement
```
Comment (otoolep) [inline CODING_GUIDELINES.md]: @corylanou -- would care to provide the main points for this section? I know it's something you care about. I am thinking about both simple swtitches on variable values and type switches.

File: CODING_GUIDELINES.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+## When -- and when not -- set a channel to 'nil'
```
Comment (otoolep) [inline CODING_GUIDELINES.md]: @jwilder -- you follow this pattern in the code. Anything you wish to add here, that I document?

File: CODING_GUIDELINES.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# Background
+
+The goal of this guide is to capture some Do and Don'ts of Go code for the InfluxDB database. When it comes to Go, writing good code is often achieved with the help of tools like `go fmt` and `go vet`. However there are still some practices not enforceable any any tools. This guide lists some specific practices to follow when writing code for the database.
```
Comment (aidylewis) [inline CODING_GUIDELINES.md]: "practices not enforceable **any any** tools" => practices not enforceable **by any** tools

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -41,13 +41,9 @@ We really like to receive feature requests, as it helps us prioritize our work.
...
 ---------------
 
-InfluxDB follows standard Go project structure. This means that all
-your go development are done in `$GOPATH/src`. GOPATH can be any
-directory under which InfluxDB and all its dependencies will be
-cloned. For more details on recommended go project's structure, see
-[How to Write Go Code](http://golang.org/doc/code.html) and
-[Go: Best Practices for Production Environments](http://peter.bourgon.org/go-in-production/), or you can just follow
-the steps below.
+InfluxDB follows standard Go project structure. This means that all your Go development are done in `$GOPATH/src`. GOPATH can be any directory under which InfluxDB and all its dependencies will be cloned. For full details on the project structure, follow along below.
```
Comment (aidylewis) [inline CONTRIBUTING.md]: "all your Go development **are** done in `$GOPATH/src`" => all your Go development **is** done in `$GOPATH/src`

File: CODING_GUIDELINES.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# Background
+
+The goal of this guide is to capture some Do and Don'ts of Go code for the InfluxDB database. When it comes to Go, writing good code is often achieved with the help of tools like `go fmt` and `go vet`. However there are still some practices not enforceable any tools. This guide lists some specific practices to follow when writing code for the database.
```
Comment (benbjohnson) [inline CODING_GUIDELINES.md]: > not enforceable *by* any tools

File: CODING_GUIDELINES.md (hunk 2 above)
Comment (benbjohnson) [inline CODING_GUIDELINES.md]: I can't think of a good time to set a channel to `nil`.

File: CODING_GUIDELINES.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        bar = "still more to do"
+        return true
+    }()
+
+    if more {
+        qux = "finished"
+    }
+    return
+```
+This allows us to use `defer` but ensures that if any new cases are added to the logic within the anonymous function, the lock will always be released.
```
Comment (benbjohnson) [inline CODING_GUIDELINES.md]: Also important to note that `defer` will still run during a panic so locks will be unlocked even if a panic occurs.

File: CODING_GUIDELINES.md (hunk 2 above)
Comment (otoolep) [inline CODING_GUIDELINES.md]: @benbjohnson -- last time I looked we did in our code. Stuff like this:

```
close(done)
wg.Wait()
done = nil
```
You saying we shouldn't be doing this? I think the `nil` state of `done` may have been used to indicate if the owning object was actually open.

Comment (benbjohnson) [inline CODING_GUIDELINES.md]: A lot of our code tries to clean up objects under the assumption that they could be reopened but I don't think that's a valid case typically. I'm leaning toward initializing objects once and letting them die hard if they try to be reopened. For example, we initialize `closing` channels in `Open()` whereas it should probably be in `NewXXX()` and we can let `Open()` panic if it tries to be called twice.

Also, owning objects should take responsibility for removing access to them when they're no longer valid so we generally shouldn't need an "is open" flag.

**Pull request merged**